File: AndroidArchitecture/mvvm-live/src/main/java/com/codearms/maoqiqi/app/addedittask/AddEditTaskFragment.java
package com.codearms.maoqiqi.app.addedittask;

import androidx.lifecycle.Observer;
import androidx.databinding.DataBindingUtil;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.codearms.maoqiqi.app.Event;
import com.codearms.maoqiqi.app.R;
import com.codearms.maoqiqi.app.base.BaseFragment;
import com.codearms.maoqiqi.app.databinding.FragmentAddEditTaskBinding;
import com.codearms.maoqiqi.app.tasks.TasksActivity;
import com.codearms.maoqiqi.app.utils.MessageMap;
import com.codearms.maoqiqi.app.utils.SnackbarUtils;

/**
 * Main UI for the add task screen. Users can enter a task title and description.
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/12 13:32
 */
public class AddEditTaskFragment extends BaseFragment {

    private AddEditTaskViewModel addEditTaskViewModel;

    public static AddEditTaskFragment newInstance() {
        return new AddEditTaskFragment();
    }

    void setViewModel(AddEditTaskViewModel addEditTaskViewModel) {
        this.addEditTaskViewModel = addEditTaskViewModel;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        FragmentAddEditTaskBinding binding = DataBindingUtil.inflate(inflater, R.layout.fragment_add_edit_task, container, false);
        binding.setAddEditTaskViewModel(addEditTaskViewModel);
        binding.setLifecycleOwner(getActivity());
        setHasOptionsMenu(true);
        return binding.getRoot();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        addEditTaskViewModel.getAddTaskEvent().observe(this, new Observer<Event<Object>>() {
            @Override
            public void onChanged(Event<Object> objectEvent) {
                if (objectEvent.getContentIfNotHandled() != null) {
                    showTasks();
                }
            }
        });
        addEditTaskViewModel.getMessage().observe(this, new Observer<Event<String>>() {
            @Override
            public void onChanged(Event<String> stringEvent) {
                String message = stringEvent.getContentIfNotHandled();
                if (message != null) showMessage(message);
            }
        });
    }

    @Override
    public void onResume() {
        super.onResume();
        addEditTaskViewModel.start();
    }

    private void showTasks() {
        if (getActivity() == null) return;
        getActivity().setResult(TasksActivity.RESULT_ADD_TASK);
        getActivity().finish();
    }

    private void showMessage(String message) {
        SnackbarUtils.show(getView(), MessageMap.get(message));
    }
}

File: AndroidArchitecture/mvvm-live-kotlin/src/main/java/com/codearms/maoqiqi/app/addedittask/AddEditTaskFragment.kt
package com.codearms.maoqiqi.app.addedittask

import androidx.lifecycle.Observer
import androidx.databinding.DataBindingUtil
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import com.codearms.maoqiqi.app.R
import com.codearms.maoqiqi.app.base.BaseFragment
import com.codearms.maoqiqi.app.databinding.FragmentAddEditTaskBinding
import com.codearms.maoqiqi.app.tasks.TasksActivity
import com.codearms.maoqiqi.app.utils.MessageMap
import com.codearms.maoqiqi.app.utils.getViewModelFactory
import com.codearms.maoqiqi.app.utils.show

/**
 * Main UI for the add task screen. Users can enter a task title and description.
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/12 13:32
 */
class AddEditTaskFragment : BaseFragment() {

    private val addEditTaskViewModel by viewModels<AddEditTaskViewModel> { getViewModelFactory() }
    // add_edit_task_fragment
//    private val args:AddEditTaskFragment by navArgs()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        val binding = DataBindingUtil.inflate<FragmentAddEditTaskBinding>(inflater, R.layout.fragment_add_edit_task, container, false)
        binding.addEditTaskViewModel = addEditTaskViewModel
        // Set the lifecycle owner to the lifecycle of the view
        binding.lifecycleOwner = this.viewLifecycleOwner
        setHasOptionsMenu(true)
        return binding.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        addEditTaskViewModel.addTaskEvent.observe(viewLifecycleOwner, Observer { objectEvent ->
            if (objectEvent!!.getContentIfNotHandled() != null) {
                showTasks()
            }
        })
        addEditTaskViewModel.message.observe(viewLifecycleOwner, Observer { stringEvent ->
            val message = stringEvent!!.getContentIfNotHandled()
            if (message != null) showMessage(message)
        })
    }

    override fun onResume() {
        super.onResume()
        addEditTaskViewModel.start("")
    }

    private fun showTasks() {
        activity?.apply {
            setResult(TasksActivity.RESULT_ADD_TASK)
            finish()
        }
    }

    private fun showMessage(message: String?) {
        view?.show(MessageMap.get(message))
    }
}